File: 0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        left_map = {"(": ")", "{": "}", "[": "]"}
        right_map = {")": "(", "}": "{", "]": "["}

        stack = []

        for p in s:
            if p in left_map:
                stack.append(p)

            if p in right_map:
                if len(stack) == 0:
                    return False
                elif left_map[stack.pop()] != p:
                    return False

        return len(stack) == 0

File: java/0020-valid-parentheses.java
import java.lang.*;
import java.util.*;

class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> openMap = Map.of('(', ')', '{', '}', '[', ']');
        Map<Character, Character> closeMap = Map.of(')', '(', '}', '{', ']', '[');

        Stack<Character> stack = new Stack<>();
        for (char c : s.toCharArray()) {
            if (openMap.containsKey(c)) {
                stack.push(c);
            } else if (closeMap.containsKey(c)) {
                if (stack.isEmpty() || openMap.get(stack.pop()) != c) return false;
          }
        }
        
        return stack.isEmpty();
    }
}